psalm/EventInterfaceChecks.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LaminasPsalm\EventManager;
+
+use Laminas\EventManager\EventInterface;
+
+class EventInterfaceChecks
+{
+    /**
+     * @param EventInterface<CheckObject, array{foo: int, bar: CheckObject}> $e
+     * @return array{
+     *     CheckObject,
+     *     array{foo: int, bar: CheckObject},
+     *     int,
+     *     CheckObject
+     * }
+     */
+    public function checkTargetAndParamsMatchTemplate(EventInterface $e): array
+    {
+        return [
+            $e->getTarget(),
+            $e->getParams(),
+            $e->getParams()['foo'],
+            $e->getParams()['bar'],
+        ];
+    }
+
+    /**
+     * @param EventInterface<null, array<empty, empty>> $e
+     * @return array{
+     *     EventInterface<CheckObject, array<empty, empty>>,
+     * }
+     */
+    public function checkSetTargetChangesTemplate(EventInterface $e): array
+    {
+        $e->setTarget(new CheckObject());
```

Comment:
One think bit me hard while improving `laminas/laminas-stdlib` here: `@psalm-self-out` removed inheritance information.

This may need hardening, but a test like this is needed, verifying that a **concrete** implementation of the event is given, and that the concrete implementation type is preserved

Reply:
For reference: https://github.com/laminas/laminas-stdlib/pull/78

Reply:
@Ocramius You were absolutely right! Looks like `@psalm-this-out` will redefine the entire type. Adding `static` won't work because the child class may use a different template setup. https://github.com/laminas/laminas-eventmanager/blob/85540cbb1da98210baba24dd23e8b479f410dbf3/psalm/Model/CheckEvent.php#L9-L15

In that case I fear we may not use `@psalm-this-out` at all. I could change it so `setTarget()` and `setParams()` only accept the template types, but at this point I'm wondering if these additions remain useful at all =S

Reply:
Yeah, it's a bit of a headscratcher :thinking: 

 * what about `&$this`? (no idea if that's legal)
 * what about assuming child classes will (for now) not have template types?